google_guest_agent/osinfo/osinfo_windows.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	begin := int(start) - int(blockStart)
+	// For translation data length is bytes.
+	trans := block[begin : begin+int(length)]
+
+	// Each 'translation' is 4 bytes long (2 16-bit sections), we just want the
+	// first one for simplicity.
+	t := make([]byte, 4)
+	// 16-bit language ID little endian
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/dd318693(v=vs.85).aspx
+	t[0], t[1] = trans[1], trans[0]
+	// 16-bit code page ID little endian
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/dd317756(v=vs.85).aspx
+	t[2], t[3] = trans[3], trans[2]
+
+	return fmt.Sprintf("%x", t), nil
+}
+
+// https://msdn.microsoft.com/en-us/library/windows/desktop/ms647464(v=vs.85).aspx
+func getStringFileInfo(block []byte, langCodePage, name string) (string, error) {
+	var start uint
+	var length uint
+	blockStart := uintptr(unsafe.Pointer(&block[0]))
+	if ret, _, _ := procVerQueryValueW.Call(
+		blockStart,
+		uintptr(unsafe.Pointer(syscall.StringToUTF16Ptr(fmt.Sprintf(`\StringFileInfo\%s\%s`, langCodePage, name)))),
+		uintptr(unsafe.Pointer(&start)),
+		uintptr(unsafe.Pointer(&length))); ret == 0 {
+		return "", errors.New("zero return code from VerQueryValueW indicates failure")
+	}
+	begin := int(start) - int(blockStart)
+	// For version information length is characters (UTF16).
+	result := block[begin : begin+int(2*length)]
+
+	// Result is UTF16LE.
+	u16s := make([]uint16, length)
+	for i := range u16s {
+		u16s[i] = uint16(result[i*2+1])<<8 | uint16(result[i*2])
+	}
+
+	return syscall.UTF16ToString(u16s), nil
+}
+
+func getVersion(block []byte, langCodePage string) (string, string, error) {
+	ver, err := getStringFileInfo(block, langCodePage, "FileVersion")
+	if err != nil {
+		return "", "", err
+	}
+	rel, err := getStringFileInfo(block, langCodePage, "ProductVersion")
+	return ver, rel, err
+}
+
+func getKernelInfo() (string, string, error) {
+	root := os.Getenv("SystemRoot")
+	if root == "" {
+		root = `C:\Windows`
+	}
+	path := filepath.Join(root, "System32", "ntoskrnl.exe")
+	if _, err := os.Stat(path); err != nil {
+		return "", "", err
+	}
+
+	pPtr := unsafe.Pointer(syscall.StringToUTF16Ptr(path))
+
+	size, _, _ := procGetFileVersionInfoSizeW.Call(
+		uintptr(pPtr))
+	if size <= 0 {
+		return "", "", errors.New("GetFileVersionInfoSize call failed, data size can not be 0")
+	}
+
+	info := make([]byte, size)
+	if ret, _, _ := procGetFileVersionInfoW.Call(
+		uintptr(pPtr),
+		0,
+		uintptr(len(info)),
+		uintptr(unsafe.Pointer(&info[0]))); ret == 0 {
+		return "", "", errors.New("zero return code from GetFileVersionInfoW indicates failure")
+	}
+
+	// This should be something like 040904b0 for US English UTF16LE.
+	langCodePage, err := getTranslation(info)
+	if err != nil {
+		return "", "", fmt.Errorf("getTranslation() error: %v", err)
+	}
+
+	return getVersion(info, langCodePage)
+}
+
+type win32OperatingSystem struct {
+	Caption, Version string
```

Comment:
Can you also describe what these fields mean in the context of the osinfo implementation for windows?